### PR TITLE
Platform should expose a timeout option

### PR DESCRIPTION
**Pull request by engyebrahim**
fix: https://github.com/microsoft/testfx/issues/3500

### CODE REVIEW

File: src/Platform/Microsoft.Testing.Platform/Resources/PlatformResources.resx [hunk 1]
```diff
@@ -602,4 +602,7 @@ Read more about Microsoft Testing Platform telemetry: https://aka.ms/testingplat
   <data name="TerminalOutputOptionInvalidArgument" xml:space="preserve">
     <value>--output expects a single parameter with value 'Normal' or 'Detailed'.</value>
   </data>
+  <data name="PlatformCommandLineTimeoutArgumentErrorMessage" xml:space="preserve">
+    <value>'timeout' option should have one argument as string in the format &lt;value&gt;[h|m|s] and 'value' is float</value>
```
Comment (Evangelink) [inline src/Platform/Microsoft.Testing.Platform/Resources/PlatformResources.resx]: Nit

```suggestion
    <value>'timeout' option should have one argument as string in the format &lt;value&gt;[h|m|s] where 'value' is float</value>
```

File: src/Platform/Microsoft.Testing.Platform/Hosts/TestHostBuilder.cs [hunk 2]
```diff
@@ -236,8 +237,30 @@ public async Task<ITestHost> BuildAsync(
...
         ICommandLineOptions commandLineOptions = serviceProvider.GetCommandLineOptions();
+
+        // setting the timeout
+        if (commandLineOptions.IsOptionSet(PlatformCommandLineProvider.TimeoutOptionKey) && commandLineOptions.TryGetOptionArgumentList(PlatformCommandLineProvider.TimeoutOptionKey, out string[]? args))
+        {
+            string arg = args[0];
+            int size = arg.Length;
+            if (!float.TryParse(arg[..(size - 1)], out float value))
+            {
+                throw new InvalidOperationException(PlatformResources.PlatformCommandLineTimeoutArgumentErrorMessage);
```
Comment (Evangelink) [inline src/Platform/Microsoft.Testing.Platform/Hosts/TestHostBuilder.cs]: We can use `throw ApplicationStateGuard.Unreachable()`, because this is already validated so this would be some unexpected state.

```suggestion
                throw ApplicationStateGuard.Unreachable();
```

File: src/Platform/Microsoft.Testing.Platform/Hosts/TestHostBuilder.cs [hunk 3]
```diff
@@ -236,8 +237,30 @@ public async Task<ITestHost> BuildAsync(
...
+            }
+
+            TimeSpan timeout = char.ToLowerInvariant(arg[size - 1]) switch
+            {
+                'h' => TimeSpan.FromHours(value),
+                'm' => TimeSpan.FromMinutes(value),
+                's' => TimeSpan.FromSeconds(value),
+                _ => throw new InvalidOperationException(PlatformResources.PlatformCommandLineTimeoutArgumentErrorMessage),
```
Comment (Evangelink) [inline src/Platform/Microsoft.Testing.Platform/Hosts/TestHostBuilder.cs]:
```suggestion
                _ => throw ApplicationStateGuard.Unreachable(),
```

File: src/Platform/Microsoft.Testing.Platform/Hosts/TestHostBuilder.cs [hunk 4]
```diff
@@ -236,8 +237,30 @@ public async Task<ITestHost> BuildAsync(
             Logging.AddProvider((_, _) => loggingState.FileLoggerProvider);
         }
 
-        // Register the server mode log forwarder if needed. We follow the console --diagnostic behavior.
```
Comment (Evangelink) [inline src/Platform/Microsoft.Testing.Platform/Hosts/TestHostBuilder.cs:239]: Why is this removed?

File: src/Platform/Microsoft.Testing.Platform/Hosts/TestHostBuilder.cs [hunk 5]
```diff
@@ -236,8 +237,30 @@ public async Task<ITestHost> BuildAsync(
...
+
+            TimeSpan timeout = char.ToLowerInvariant(arg[size - 1]) switch
+            {
+                'h' => TimeSpan.FromHours(value),
+                'm' => TimeSpan.FromMinutes(value),
+                's' => TimeSpan.FromSeconds(value),
+                _ => throw new InvalidOperationException(PlatformResources.PlatformCommandLineTimeoutArgumentErrorMessage),
+            };
+
+            testApplicationCancellationTokenSource.CancelAfter(timeout);
```
Comment (Evangelink) [inline src/Platform/Microsoft.Testing.Platform/Hosts/TestHostBuilder.cs:259]: I think that's fine as-is but I'd like to get @MarcoRossignoli's view. Marco, shall we do that directly when initializing the test app source object?

**Review by Evangelink — COMMENTED** [review]
Some test and it's perfect!

File: src/Platform/Microsoft.Testing.Platform/Hosts/TestHostBuilder.cs (hunk 4 above)
Comment (engyebrahim) [reply src/Platform/Microsoft.Testing.Platform/Hosts/TestHostBuilder.cs:239]: it's still there but after the code i added

File: src/Platform/Microsoft.Testing.Platform/Hosts/TestHostBuilder.cs (hunk 5 above)
Comment (MarcoRossignoli) [reply src/Platform/Microsoft.Testing.Platform/Hosts/TestHostBuilder.cs:259]: I think it's fine we build the appct before the command line. And I prefer to use the `ICommandLineOptions` where timing allows it.

File: test/IntegrationTests/Microsoft.Testing.Platform.Acceptance.IntegrationTests/TimeoutTests.cs [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        testHostResult.StandardError.Contains("'timeout' option should have one argument as string in the format <value>[h|m|s] where 'value' is float");
+    }
+
+    [ArgumentsProvider(nameof(TargetFrameworks.All), typeof(TargetFrameworks))]
+    public async Task TimeoutWithValidArg_WithTestTimeOut_BuildFails(string tfm)
+    {
+        var testHost = TestInfrastructure.TestHost.LocateFrom(_testAssetFixture.NoExtensionTargetAssetPath, TestAssetFixture.AssetName, tfm);
+        TestHostResult testHostResult = await testHost.ExecuteAsync("--timeout 1s");
+
+        testHostResult.AssertExitCodeIsNot(ExitCodes.Success);
```
Comment (Evangelink) [inline test/IntegrationTests/Microsoft.Testing.Platform.Acceptance.IntegrationTests/TimeoutTests.cs:53]: Let's assert we see the cancelling message.

File: test/IntegrationTests/Microsoft.Testing.Platform.Acceptance.IntegrationTests/TimeoutTests.cs [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    [ArgumentsProvider(nameof(TargetFrameworks.All), typeof(TargetFrameworks))]
+    public async Task TimeoutWithValidArg_WithHourAsSuffix_WithTestNotTimeOut_BuildPasses(string tfm)
+    {
+        var testHost = TestInfrastructure.TestHost.LocateFrom(_testAssetFixture.NoExtensionTargetAssetPath, TestAssetFixture.AssetName, tfm);
+        TestHostResult testHostResult = await testHost.ExecuteAsync("--timeout 1h");
+
+        testHostResult.AssertExitCodeIs(ExitCodes.Success);
+    }
```
Comment (Evangelink) [inline test/IntegrationTests/Microsoft.Testing.Platform.Acceptance.IntegrationTests/TimeoutTests.cs]: All these tests happens after the execution is done. Maybe update the name so that's more explicit and we can also check the output doesn't contain the cancel message.
